### ISSUE TITLE
Improved MAF metrics for survey uniformity

### ISSUE THREAD

**Issue by rmandelb** [LSSTDESC/SprintWeek2023#22]
# Improved MAF metrics for survey uniformity

Following discussions in the OS working group (#desc-obs-strat and #desc-wfd-uniformity-and-rolling), we'd like to produce improved MAF metrics quantifying survey uniformity.  These should be tested and demonstrated on the latest OpSim runs.

Contacts: Rachel Mandelbaum
Day/Time: TBD depending on participant availability
Main communication channel: #desc-wfd-uniformity-and-rolling
GitHub repo: https://github.com/lsst/rubin_sim
Zoom room (if applicable): TBD

## Goals and deliverable

We'd like a set of improved MAF metrics along with documentation describing their use and the interpretation of their outputs.

## Resources and skills needed

At least one person familiar with MAF and how to use OpSim runs, but those with a general interest in survey uniformity are welcome to participate!

## Detailed description

TBD

**Comment by reneehlozek** [comment]
So excited to sprint this!

**Comment by beckermr** [comment]
OK here are some notes from a small discussion that @reneehlozek and I had on some planning:

## identified todos

 - [x] @beckermr to send along code for plots
 - [ ] code up slightly more sophisticated depth metric
 - [ ] six month delay in the scheduler (@beckermr to code up)


## possible metrics beyond depth

 - [ ] Measure change in mask area and/or structure as a function of redshift (We've seen interesting effects in DES based on this and it impacts the covariance matrix.)
 - [ ] variation in mean redshift as a function position and related to w(theta) using results from this paper (https://arxiv.org/abs/2305.15406)
 - [ ] use SSI results from Boris++ to motivate a simple model mapping survey properties to variations in density of objects, and then estimate impact on clustering
 - [ ] derive a specific metric for how uniform year 1 needs to be in order to start rolling (@erykoff)
 - [ ] figure out scheme to add noise to the deepest cell coadds in the data to reduce variations in depth before we run shear codes, etc.

**Comment by rmandelb** [comment]
Thanks, @beckermr .

Following a discussion on Slack, we'll have an initial discussion amongst those participating in this sprint **right after lunch on Monday (around 1pm Eastern)**.  Location and Zoom coordinates are TBD.

**Comment by rmandelb** [comment]
I took the liberty of reserving a room for us each day next week at 10am Pacific / 1pm Eastern (Scott Hall 6142 - see signups [here](https://confluence.slac.stanford.edu/pages/viewpage.action?pageId=388856811)) so we have a dedicated time period with a quiet room for us to have Zoom discussions with remote participants.  This was just an initial guess at what we might need, and we can adjust this plan during our first tagup.  My thinking was that we'll use this time to plan out work, then go off and work without necessarily being on Zoom together.  Other suggestions for how to manage the hybrid sprint are welcome.

**Comment by beckermr** [comment]
@reneehlozek here is the code I promised: https://github.com/beckermr/misc/tree/main/work/rubin_rolling_uniformity

**Comment by rmandelb** [comment]
I made an [intro slide](https://docs.google.com/presentation/d/12MMp1c_USgXcXTIxzJXWo0me3I_FYHMz_-R8tBLS1TQ/edit#slide=id.g28cff4e86bc_0_0) for the morning tagup on day 1; anybody involved is welcome to edit it.  I semi-arbitrarily listed two of us who will be there in person as contacts.

**Comment by rmandelb** [comment]
Discussion in ~10 mins on Zoom

**Comment by reneehlozek** [comment]
Tasks:

**Mean z** @rmandelb @reneehlozek @arunkannawadi 

- [ ] Compute the N(z) from the i-band magnitude from sims
- [ ] Develop tomographic binning strategy
- [ ] Use that strategy to determine the meanz per pixel within the bin
- [ ] Push the meanz to Clkk error (either using [2305.15406](2305.15406) or otherwise)
- [ ] Push the Clkk error to cosmology error
- [ ] Think through how this is communicated outside of static science folks


**SSI** @beckermr @egawiser Federico @sidneymau 

- [ ] Speak to Boris/get code from Boris re: SSI code
- [ ] Compute the change in density within some z bin in e.g. quadrants on the sky
- [ ] Run Humna's previous code on new sims (see [1605.00555](https://arxiv.org/abs/1605.00555))
- [ ] Use code from Humna's LSS SCOC recommendations to push those changes to gal density (compare the SSI results to Humna's code and replace density multipliers with SSI results)
- [ ] Back s8 uncertainties to uncertainty on the survey density
- [ ] Can we get this uncertainty out of the SSI results?
- [ ] Think through how this is communicated outside of static science folks

**For both**

- [ ] Compare against other DESC metrics 
- [ ] Communicate to OS WG

**Comment by rmandelb** [comment]
@erykoff - Example MAF metric that does lots of non-MAF-y things: https://github.com/lsst/rubin_sim/blob/62f72605d3f1a493aadf74e364efa6af462f74e1/rubin_sim/maf/maf_contrib/static_probes_fom_summary_metric.py#L12

**Comment by reneehlozek** [comment]
[Note from Humna](https://docs.google.com/document/d/1rQB76P_B1ONVSmnHYKXFNM1oB5IgdPD6aJ4qU-6kEnk/edit) on how to get rubin_sim running using the DESC NERSC environment

**Comment by rmandelb** [comment]
@reneehlozek @arunkannawadi - in the DESC SRD https://arxiv.org/pdf/1809.01669.pdf : 
- Appendix F1 gives photometric (i.e., clustering) sample number density as a function of i-band limiting mag in equation 9.
- Figure F2 shows the n(z) for two particular limiting magnitudes.  So that gives two data points for (limiting magnitude, `<z>`).
- We can use the relationship between depth and the `<z>` of the sample from the LSST science book as a functional form to interpolate between those two data points.

**Comment by rmandelb** [comment]
Following a discussion with @aimalz  -

If we are trying to split the mean-z part of this into two, namely "OpSim output to shifts in n(z) per tomo bin in each pixel" and then "shifts in n(z) per tomo bin in each pixel to some cosmological quantity", then Ellen Hang might have done a lot of work that will take care of the first part.
- If we have a mock catalog like SkySim, her code in https://github.com/LSSTDESC/rail_astro_tools/blob/main/src/rail/creation/degradation/observing_condition_degrader.py will apply magnitude errors based on the observing strategy in each pixel (i.e., based on an OpSim output).
- Then we can run a photo-z code in RAIL.
- Then the tomographic binning routines like https://github.com/LSSTDESC/rail_base/blob/main/src/rail/estimation/algos/equal_count.py can be used to assign galaxies to tomographic bins.
- Then we can easily know what happens to the n(z) across the survey

This is more sophisticated than the "fake it with a parametric n(z) and some approximation to what might happen in each tomographic bin".  We wouldn't put this into MAF, but rather could use this to determine some quantities that will go into MAF.

My homework: ping Ellen to find out about an example notebook illustrating these routines, because there is enough existing work that we might be able to do this during this week.   Edited to add link to notebook: https://github.com/LSSTDESC/rail/blob/main/examples/creation_examples/example_ObsConditions.ipynb

Work on the second step of the process can of course proceed in parallel.

**Comment by rmandelb** [comment]
Hi all - for today, we will tag up again at 1pm Eastern / 10am Pacific.  **For in-person participants:** because of multiple tutorials, we will be in 7423 Wean Hall.  I can guide any of the non-CMU folks.

**Comment by rmandelb** [comment]
@hangqianjun - one important question that will affect how we can use this example: it seems to rely on some stored outputs in RAIL (for example, "examples_data/creation_data/data/survey_conditions/minion_1016_dc2_Median_airmass_i_and_nightlt1825_HEAL.fits") rather than interfacing with `rubin_sims`.  Am I reading this correctly?

We need code that can run on new OpSim outputs, so the code has to be able to use the `rubin_sims` package and talk to the database outputs.  How was this example file in RAIL generated?  It must involve some code that uses the OpSim outputs, and that code would be needed for us to use this on any new observing strategies.  Can you please point us to it?  And does your degrader let you pass in something that says what strategy to use?  Thanks!

**Comment by hangqianjun** [comment]
@rmandelb Yes, the input maps come from rubin_sims in .npz format and I have converted them into healpix for generality of the input. Interfacing with rubin_sims would mean to have rail to depend on rubin_sims, I worry that it might be too heavy for `rail_astro_tools` (@aimalz might know better). Alternatively we could have a separate `rubin-rail` interface that integrates basic rail into rubin?
Ultimately if we just want to run them in series, then it is just a one line code to convert the rubin_sim outputs into healpix inputs:
```
from rubin_sim import maf
import healpy as hp

def load_maf_map(fname):
    
    fin = maf.MetricBundle.load(fname)
    #counts/map values
    opmap=fin.metricValues
    mafmap=np.copy(opmap.data)
    mafmap[opmap.mask]=np.nan
    
    return mafmap

def maf_maps_to_fits(fname_in):
    usemap = load_maf_map(fname_in)
    # save this into fits file:
    fname = fname_in[:-3] + 'fits'
    hp.write_map(fname, usemap)
    print("Written: %s"%fname)
```

**Comment by rmandelb** [comment]
Hi - What i had in mind was running them in series rather than making explicit dependencies - I need code that can do one task after the other, but I was missing the glue for those two tasks.  This is very helpful, thanks!

**Comment by reneehlozek** [comment]
**Issues from the Oct 17 Tag up**

- [x] Write simple new metric based on coadd5m and test the bundling/running (look at WeaklensingNvisits) -- crucially we want a metric that doesn't operate on just one pixel, but uses the whole map to output a number at one pixel @erykoff and @reneehlozek will tag up on this 
- [x] Compute the function to map m5mag to mean z in tomographic bins based on dividing parametric N(z) into quintiles @arunkannawadi 
- [ ] Incorporate the RAIL functions from @hangqianjun into a notebook to run on the latest opsim run @rmandelb 
- [ ] get links to previous code from Humna and try and re-run on new opsims @egawiser

**Comment by arunkannawadi** [comment]
Here's a notebook with the relevant code to map m5mag (what I call `ilim`) to mean z and its sensitivity (which is constant per tomographic bin per our assumptions): https://gist.github.com/arunkannawadi/7d1cf0c8f4105f6deed435428e687643

**Comment by beckermr** [comment]
Can you double check this against DES y3?

**Comment by rmandelb** [comment]
Thanks for making a start at this!  I didn't check the code in detail, but from the results, something is going wrong... We know from the start that <z> for the overall Y10 sample is 1.08 (that's one boundary condition).  But the code in cell 9 is saying that for our 5 tomo bins, the means are [0.07674694118450194, 0.13956359371383936, 0.19563589335277862, 0.26564695816129164,0.41695206813304336].  Those two statements are mutually inconsistent -- all 5 bins cannot have mean redshifts well below the overall sample mean.  So I think there is a bug somewhere.

**Comment by arunkannawadi** [comment]
Great catch, Rachel. The missing component here is the proper normalization when integrating within a redshift bin. I'll update the notebook with that missing factor tomorrow. Edit: the numbers are smaller by a factor that equals the number of tomographic bins.

**Comment by reneehlozek** [comment]
Following a chat with Lynne on slack I think the consensus was for @erykoff to write a metric using the maf output (e.g. Nvisits) but not try and integrate that into maf himself - this is something to pass on to the team afterwards.

**Comment by rmandelb** [comment]
Joining Zoom now in 6142 - will be on in a minute

**Comment by reneehlozek** [comment]
**Notes from Wed 18 tag up**

- [x] Integrate the two-plot code into the other metric notebook @rmandelb 
- [x] Push code to repo and test output @arunkannawadi 
- [x] Make time-slicing plots for metrics @reneehlozek 
- [ ] Get code from Boris/Humna for deltadelta @egawiser 
- [x] Find out why code is taking too long and get gorgeous plots @erykoff

**Comment by rmandelb** [comment]
**Notes from Thurs Oct 19 tag up**

- Federico: connecting uncertainties in OS overdensity to uncertainties in S8 (sweeping galaxy bias under the rug, assuming just fitting to sigma8), still working on this.  If we can get Humna's code to translate OS overdensity to LSS overdensity that would give us a reasonable result.  Can tie to i_lim variations -> number density fluctuations.  Will need to connect to how well we expect to be able to measure number density fluctuations.
- Rachel: RMS i_lim variations as a function of time for various strategies.  Ask for clarity on the rolling sim variations, check with OS how they do for other things.  Propagate results to mean redshift and number density.
- Matt: rebaseline sprint plans - 6 month shift, SSI
- Eric: will work out the riz SNR for dim point-source with some red spectrum as proxy for the total depth that we want
- Arun: updated results for <z> sensitivity on i_lim with tomographic binning, will update and push notebook
- Eli: photometric calibration metric running on v3.3 baseline and noroll at NERSC, should try to understand behavior, may need to use synthetic stars.  Will do year 1, year 2, and year1+2 solutions.  Expect year 1 to be same for baseline and noroll, year 2 should look bad for baseline, year1+2 should be the same for baseline and noroll.

**Comment by rmandelb** [comment]
Hey folks - I did some sleuthing to figure out which of the modified strategies were better than baseline (since so many lines overlapped) and the answer is 3 of them:
mid1 - season6
mid2 - season4
mid2 - season6

So we need to find out what differs in these compared to baseline and to mid1 - season(2, 4) and mid2 - season 2.

**Comment by rmandelb** [comment]
To be clear, those 3 are basically identical to each other in years <=4 (with minor variations in 5+) so that's why it was hard to distinguish.

**Comment by beckermr** [comment]
Well it looks like we found a bug in the sims so I would disregard the result above. Hopefully it gets fixed and we'll rerun things.

**Comment by rmandelb** [comment]
Yeah, in any case the point is more to get the MAF metrics and analysis codes we need in place rather than to focus on these specific sims, so I think it's fine - we can keep going...

**Comment by rmandelb** [comment]
We made a fantastic start on all this work and **future work on this issue** will continue in the desc-wfd-uniformity-and-rolling Slack channel!

Summary of work undertaken, findings, and next steps for four parallel threads:

### 1. Photometric calibration metrics _(Eli)_

- Eli worked on photometric calibration metrics, which are inherently non-local.
- He has a preliminary metric coded up and was testing it.
- **Future work:** keep testing it and then get it integrated into MAF (make a PR and work with Lynne, Peter)

### 2. Potential variants in rolling strategies _(Matt)_

- Matt developed some code for variations in rolling strategies and sent it to Peter for integration/testing.
- **Future work:** work with Peter to get this going in some test runs.

### 3. Quantifying the impact of inhomogeneity of coadds on the mean sample redshift and from there to cosmology _(Renee, Rachel, Arun)_

- Arun produced a notebook that gets dmeanz/d(i-band limiting mag) under some simple assumptions.
- Renee and Rachel produced a notebook that explores variations in extinction-corrected i-band limiting magnitude across the survey, including simple quantification e.g. through standard deviations.
- This notebook was used to explore the v3.3 runs, and we found what may be a bug in some of the rolling variants (reported to Peter).
- We observed also that even for a similar standard deviation of i-band depth, there are still spatial patterns in some rolling runs that aren't present in no-roll?.  Do we need a power spectrum-based metric?
- **Future work:** Get a combined riz limiting mag (Eric?) rather than just i; clean up the above-mentioned Renee/Rachel notebook (Rachel) so it is readable and makes sense; connect it to Arun's notebook (Rachel or Arun?); connect to cosmology via that paper Matt found (Renee and/or ?); explore power spectrum-based metric?

### 4. Quantifying the impact of inhomogeneity of coadds on the overdensity (i.e., systematic uncertainty in the overdensities due to obs strat) _(Federico, Eric)_

- Federico developed a notebook with some basic quantification, but need feedback from Humna's past work to figure out which curves are realistic.
- Federico has a dataframe from Rachel with stddev(Ngal) across the survey that could be used to explore excess fluctuations due to the noroll strategy.
- What is the next step here?

**Comment by egawiser** [comment]
The next step for #4 above is to estimate an uncertainty in \delta_OS .  Federico implemented "Method 1", which compares the fractional rms in a map of N_gal to that in a depth map to get a muliplier of 0.05.  Method 2 will be to use the N_gal vs. depth relationship from the LSST Science book, but this may have already been used in Humna's N_gal routine so may not give a sufficient estimate of the uncertainty.  Method 3 would be to use Boris' SSI results to get a second opinion on the multiplier.  Once we have two sufficiently independent methods, we'll have \Delta \delta_OS and can propagate that through to see if we are below the 2% uncertainty on sigma_8 needed to avoid worsening our cosmological constraints.

**Comment by rmandelb** [comment]
10/26 tagup summary:
- We'll meet weekly (updated on DESC calendar) to follow up
- Rachel will check w/ Peter about timeline for updated sims given 3.3 bugs, also whether baseline/noroll may be affected, to get an update for Eli  _(pinged in thread in sims Slack channel)_
- Eli will make a PR in MAF in the next week, talk w/ Peter about auxiliary data
- Rachel will do PS + cleanup piece
- Renee will take the dataframe with stddev i-band limiting mag and combine with Arun's dz/dmag
- Federico will continue with the path outlined above (Eric's comments - method 2 but with DESC SRD n(ilim), then method 3 if possible)
- writeup for task force

**Comment by reneehlozek** [comment]
This is a quick note that I made a modules file in [ObsStrat](https://github.com/LSSTDESC/ObsStrat/tree/meanz_uniformity_maf) that contains the functions we have been using. I added in Arun's functions as well as Rachels/mine and an ipython notebook that now also computes the meanz (saving only the first tomographic bin for now, will update that later this week)

**Comment by rmandelb** [comment]
Woohoo, thanks Renee!  In the meantime, on Friday and Monday I did a lot of cleanup in the `iband_lim.ipynb` -- commenting, removing garbage, functionizing.  So I think to merge these, we need to:
1. Check that my cleanup did not make function defs inconsistent from yours, and replace my function defs with calls to your functions.
2. Move some of my new function definitions into your file.

I can do that in the next couple days unless you beat me to it.

**Comment by reneehlozek** [comment]
Just a brief note @rmandelb that I'm working on this now in a mini sprint before our tag up at 11

**Comment by rmandelb** [comment]
- Renee @reneehlozek  took the `iband_lim` notebook and consolidated functions from the separate module and incorporated meanz; Rachel should pull Renee's stuff and can keep plotting iband lim things while Renee cleans up the meanz plotting
- Eli @erykoff  asked Peter about simulated star catalogs (extending above dec=0) which could be important for metrics; since this would be challenging, Eli is pursuing an alternate approach to a Gaia-based star catalog following that discussion.  This will go into self-calibration metric.  Working with Jeff Carlin, Elana Urbach.
- Federico @FedericoBerlfein will provide an update on Slack.
- Matt @beckermr is iterating with Peter on strategies with phase shifts, updates are on Slack, looking promising.

**Comment by rmandelb** [comment]
- Renee has suffered from the vagaries of pandas dataframes while trying to show meanz standard deviation across the survey by strategy / year but will soldier bravely onwards.  Once meanz stuff is ready, can hand over to Rachel for cosmology connection.
- Federico is working on Delta delta, waiting to hear from Boris.  In the meantime, he will work with Rachel on using depth maps to get statistics of delta_OS.
- Rachel has updated routines to exclude twilight exposures and started looking at power spectra to quantify spatial patterns in residuals.  Will check about updating MAF metric that already exists vs. making new one, and get feedback from Matt.

**Comment by rmandelb** [comment]
11/16 update (@erykoff please add/correct as needed):
- Eli @erykoff expects to have the new reference catalog, to be used for a great many things, the week after Thanksgiving.  He can then incorporate a downsampled version into his metric and make a PR.  We will have to decide what sims to run it on because it's computationally intensive.  A possible default would be to always have it run on the first few years of any new baseline sim, and then run it only on selected others if we have a particular reason to consider them.
- I've been slow due to other things but will figure out whether to fork or branch from `rubin_sim` and then can incorporate the power spectrum-based metric by modifying theirs to include what I need (monopole subtraction).
- We discussed the connection between the meanz metric and cosmology based on figure 9 of https://arxiv.org/pdf/2305.15406.pdf and I have a plan for that; will illustrate with a notebook.
